src/hotspot/share/gc/shenandoah/shenandoahTaskqueue.hpp

```diff
@@ -143,22 +143,28 @@ class ObjArrayChunkedTask
     chunk_shift  = pow_shift + pow_bits
   };
 
+  static const uintptr_t oop_decode_mask = right_n_bits(oop_bits) - 3;
+  // When lowest bit is set, don't count liveness.
+  static const uintptr_t count_liveness_decode_mask = 1;
+  // When 2nd bit is set, mark as finalizable
+  static const uintptr_t strong_decode_mask = 1 << 1;
+
 public:
-  ObjArrayChunkedTask(oop o = NULL) {
-    assert(decode_oop(encode_oop(o)) ==  o, "oop can be encoded: " PTR_FORMAT, p2i(o));
-    _obj = encode_oop(o);
+  ObjArrayChunkedTask(oop o = NULL, bool count_liveness = true, bool strong = true) {
+    assert(decode_oop(encode_oop(o, count_liveness, strong)) ==  o, "oop can be encoded: " PTR_FORMAT, p2i(o));
```

Comment:
Need encodeability `assert`-s for `count_liveness` and `strong` too? Also in the other constructor? This might get costly, and the whole thing might need rethinking... But at least the initial testing better check there are no bugs here.